bin/livy-server

```diff
@@ -18,34 +18,150 @@
 #
 # Runs Livy server.
 
-set -e
+
+usage="Usage: livy-server (start|stop)"
+
+# Find the java binary
+if [ -n "${JAVA_HOME}" ]; then
+  RUNNER="${JAVA_HOME}/bin/java"
+else
+  if [ `command -v java` ]; then
+    RUNNER="java"
+  else
+    echo "JAVA_HOME is not set" >&2
+    exit 1
+  fi
+fi
 
 export LIVY_HOME=$(cd $(dirname $0)/.. && pwd)
-cd $LIVY_HOME
+export LIVY_CONF_DIR="${LIVY_CONF_DIR:-"${LIVY_HOME}/conf"}"
 
-LIBDIR="$LIVY_HOME/jars"
-if [ ! -d "$LIBDIR" ]; then
-  LIBDIR="$LIVY_HOME/server/target/jars"
+if [ -f "${LIVY_CONF_DIR}/livy-env.sh" ]; then
+  # Promote all variable declarations to environment (exported) variables
+  set -a
+  . "${LIVY_CONF_DIR}/livy-env.sh"
+  set +a
 fi
-if [ ! -d "$LIBDIR" ]; then
-  echo "Could not find Livy jars directory." 1>&2
-  exit 1
+
+if [ "$LIVY_PID_DIR" = "" ]; then
+  LIVY_PID_DIR=/tmp
 fi
+pid="$LIVY_PID_DIR/livy-$LIVY_IDENT_STRING-server.pid"
 
+livy_rotate_log ()
+{
+    log=$1;
+    num=5;
+    if [ -n "$2" ]; then
+	num=$2
+    fi
+    if [ -f "$log" ]; then # rotate logs
+	while [ $num -gt 1 ]; do
+	    prev=`expr $num - 1`
+	    [ -f "$log.$prev" ] && mv "$log.$prev" "$log.$num"
+	    num=$prev
+	done
+	mv "$log" "$log.$num";
+    fi
+}
 
-LIVY_CLASSPATH="$LIBDIR/*:$LIVY_HOME/conf"
+start_livy_server() {
 
-if [ -n "$SPARK_CONF_DIR" ]; then
-  LIVY_CLASSPATH="$LIVY_CLASSPATH:$SPARK_CONF_DIR"
-fi
-if [ -n "$HADOOP_CONF_DIR" ]; then
-  LIVY_CLASSPATH="$LIVY_CLASSPATH:$HADOOP_CONF_DIR"
-fi
-if [ -n "$YARN_CONF_DIR" ]; then
-  LIVY_CLASSPATH="$LIVY_CLASSPATH:$YARN_CONF_DIR"
-fi
+  LIBDIR="$LIVY_HOME/jars"
+  if [ ! -d "$LIBDIR" ]; then
+    LIBDIR="$LIVY_HOME/server/target/jars"
+  fi
+  if [ ! -d "$LIBDIR" ]; then
+    echo "Could not find Livy jars directory." 1>&2
+    exit 1
+  fi
+
+  if [ "$LIVY_IDENT_STRING" = "" ]; then
+    export LIVY_IDENT_STRING="$USER"
+  fi
+
+  # get log directory
+  if [ "$LIVY_LOG_DIR" = "" ]; then
+    export LIVY_LOG_DIR="${LIVY_HOME}/logs"
+  fi
+  mkdir -p "$LIVY_LOG_DIR"
+  touch "$LIVY_LOG_DIR"/.livy_test > /dev/null 2>&1
+  TEST_LOG_DIR=$?
+  if [ "${TEST_LOG_DIR}" = "0" ]; then
+    rm -f "$LIVY_LOG_DIR"/.livy_test
+  else
+    echo "$LIVY_IDENT_STRING don't have permission to write log directory: $LIVY_LOG_DIR"
+    exit 1
+  fi
+  log="$LIVY_LOG_DIR/livy-$LIVY_IDENT_STRING-server.out"
+
+  # Set default scheduling priority
+  if [ "$LIVY_NICENESS" = "" ]; then
+      export LIVY_NICENESS=0
+  fi
+
+  mkdir -p "$LIVY_PID_DIR"
+
+  if [ -f "$pid" ]; then
+    TARGET_ID="$(cat "$pid")"
+    if [[ $(ps -p "$TARGET_ID" -o comm=) =~ "java" ]]; then
+      echo "livy_server running as process $TARGET_ID.  Stop it first."
+      exit 1
+    fi
+  fi
+
+  livy_rotate_log "$log"
+
+  CLASSPATH="$LIBDIR/*:$LIVY_HOME/conf:$CLASSPATH"
+  command="$RUNNER $LIVY_SERVER_JAVA_OPTS -cp $CLASSPATH com.cloudera.livy.server.LivyServer"
+  if [ -z "$old_livy_launch" ]; then
```

Comment:
Thanks for adding this back, but I don't think this is the best way to do it. Mainly because in the "old_livy_launch" case you're running a whole lot of unnecessary code to deal with log rotation and pid files, when in that case all you need is to just run the command.

You could place the code that generates the `command` string in a separate function to reuse it; then just call `exec $(function_that_creates_command)` in the old case, or `nohup ... $(function_that_creates_command) > blah` in the new case.